File: src/agents/shared/fingerprint.ts
import crypto from 'crypto';
import type { NormalizedTestResult } from './types';

export function createFailureFingerprint(test: NormalizedTestResult): string {
  const normalizedError = (test.errorMessage ?? '')
    .replace(/\d+/g, '<num>')
    .replace(/\s+/g, ' ')
    .slice(0, 500);

  const raw = [
    test.testId,
    test.feature,
    normalizedError,
    test.file
  ].join('|');

  return crypto.createHash('sha256').update(raw).digest('hex');
}

File: python_agents/agents/shared/fingerprint.py
from __future__ import annotations

import hashlib
import re

from .models import NormalizedTestResult


def create_failure_fingerprint(result: NormalizedTestResult) -> str:
    error = result.error_message or ""
    normalized_error = re.sub(r"\d+", "<num>", error)
    normalized_error = re.sub(r"\s+", " ", normalized_error)[:500]
    raw = "|".join([
        result.test_id,
        result.feature,
        normalized_error,
        result.file,
    ])
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()
